libs/actions/common/PythonRunner/pythonrunner.py

```diff
@@ -0,0 +1,180 @@
+        return packages
+
+    @classmethod
+    def _get_venv_paths(cls, venv_dir: str) -> tuple[str, str]:
+        """
+        Get the paths to the python and pip executables in the virtual environment.
+
+        :param venv_dir: [str] The path to the virtual environment directory.
+        :return: [tuple] The paths to the python and pip executables.
+        """
+
+        bin_folder = "Scripts" if sys.platform == "win32" else "bin"
+        venv_bin = os.path.join(venv_dir, bin_folder)
+        venv_python = os.path.join(venv_bin, "python")
+        venv_pip = os.path.join(venv_bin, "pip")
+        return venv_python, venv_pip
+
+    @classmethod
+    def _install_packages(cls, venv_pip: str, packages: list) -> None:
+        """
+        Install the provided packages in the virtual environment.
+
+        :param venv_pip: [str] The path to the pip executable in the virtual environment.
+        :param packages: [list] The list of package names to install.
+        """
+        if packages:
+            command = [venv_pip, "install"] + packages
+            try:
+                subprocess.run(command, check=True)
+            except subprocess.CalledProcessError as e:
+                logger.error(f"Error installing packages: {e}")
+
+    @classmethod
+    def _check_and_install_virtualenv(cls) -> None:
+        """
+        Check if virtualenv is installed, if not, install it.
+        """
+        try:
+            import virtualenv
+
+            logger.info("virtualenv is already installed.")
+        except ImportError:
+            logger.info("virtualenv is not installed, installing now...")
+            subprocess.check_call(
+                [sys.executable, "-m", "pip", "install", "virtualenv"]
+            )
+
+    @classmethod
+    def check_python_file_or_folder(cls, path: str, venv_pip: str) -> str:
+        if os.path.isdir(path):
+            requirements_path = os.path.join(path, "requirements.txt")
+            main_script_path = os.path.join(path, "main.py")
+            if os.path.exists(requirements_path):
+                with open(requirements_path, "r") as file:
+                    packages = file.read().splitlines()
+                PythonRunner._install_packages(venv_pip, packages)
+                if os.path.exists(main_script_path):
+                    return main_script_path
+                else:
+                    logger.error(f"main.py not found in the directory {path}")
+            else:
+                logger.error(f"No requirements.txt found in the directory {path}")
+        elif os.path.isfile(path):
+            packages = PythonRunner._get_module_names_from_script(path)
+            PythonRunner._install_packages(venv_pip, list(packages))
+            if os.path.exists(path):
+                return path
+        else:
+            logger.error("The provided path is neither a file nor a directory.")
+
+    @staticmethod
+    def create_venv_and_do_something(func: callable):
```

Comment:
@pisces860225 I would prefer to rename this function name to: `execute_with_venv`.

Reply:
This part was code left over from debugging and has been removed.

Reply:
The function name has been changed to execute_with_venv.